include/pybind11/subinterpreter.h

```diff
@@ -0,0 +1,320 @@
+
+        // Get the internals pointer (without creating it if it doesn't exist).  It's possible
+        // for the internals to be created during Py_EndInterpreter() (e.g. if a py::capsule
+        // calls `get_internals()` during destruction), so we get the pointer-pointer here and
+        // check it after.
+        auto *&internals_ptr_ptr = detail::get_internals_pp<detail::internals>();
+        auto *&local_internals_ptr_ptr = detail::get_internals_pp<detail::local_internals>();
+        {
+            dict sd = state_dict();
+            internals_ptr_ptr
+                = detail::get_internals_pp_from_capsule_in_state_dict<detail::internals>(
+                    sd, PYBIND11_INTERNALS_ID);
+            local_internals_ptr_ptr
+                = detail::get_internals_pp_from_capsule_in_state_dict<detail::local_internals>(
+                    sd, detail::get_local_internals_id());
+        }
+
+        // End it
+        Py_EndInterpreter(destroy_tstate);
+
+        // do NOT decrease detail::get_num_interpreters_seen, because it can never decrease
+        // while other threads are running...
+
+        if (internals_ptr_ptr) {
+            internals_ptr_ptr->reset();
+        }
+        if (local_internals_ptr_ptr) {
+            local_internals_ptr_ptr->reset();
+        }
+
+        // switch back to the old tstate and old GIL (if there was one)
+        if (switch_back)
+            PyThreadState_Swap(old_tstate);
+    }
+
+    /// Get a handle to the main interpreter that can be used with subinterpreter_scoped_activate
+    /// Note that destructing the handle is a noop, the main interpreter can only be ended by
+    /// py::finalize_interpreter()
+    static subinterpreter main() {
+        subinterpreter m;
+        m.istate_ = PyInterpreterState_Main();
+        m.disarm(); // make destruct a noop
+        return m;
+    }
+
+    /// Get a non-owning wrapper of the currently active interpreter (if any)
+    static subinterpreter current() {
+        subinterpreter c;
+        c.istate_ = detail::get_interpreter_state_unchecked();
+        c.disarm(); // make destruct a noop, we don't own this...
+        return c;
+    }
+
+    /// Get the numerical identifier for the sub-interpreter
+    int64_t id() const {
+        if (istate_ != nullptr)
+            return PyInterpreterState_GetID(istate_);
+        else
+            return -1; // CPython uses one-up numbers from 0, so negative should be safe to return
+                       // here.
+    }
+
+    /// Get the interpreter's state dict.  This interpreter's GIL must be held before calling!
+    dict state_dict() { return reinterpret_borrow<dict>(PyInterpreterState_GetDict(istate_)); }
+
+    /// abandon cleanup of this subinterpreter (leak it). this might be needed during
+    /// finalization...
+    void disarm() { creation_tstate_ = nullptr; }
+
+    /// An empty wrapper cannot be activated
+    bool empty() const { return istate_ == nullptr; }
```

Comment:
I forget, do we have a macro for nodiscard?

Reply:
`git grep nodiscard` doesn't reveal anything. That should be in the macro def, so no.